Ensure everything is sorted whenever a command that generates po / pot files is used

johnzhou721 opened the pull request:
Exactly what it says on the tin.

<!--- Describe your changes in detail -->
<!--- What problem does this change solve? -->
<!--- If this PR relates to an issue, include Refs #XXX or Fixes #XXX -->

## PR Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] All new features have been tested
- [x] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

johnzhou721 commented on lektor_i18n.py:
```diff
@@ -183,6 +183,7 @@ def _msg_init(self):
             "-o",
             self.FILENAME_PATTERN.format(self.language),
             "--no-translator",
+            "-F",
```
```suggestion
```

johnzhou721 replied:
this is nonexistent, sorry

johnzhou721 commented on lektor_i18n.py:
```diff
@@ -140,7 +140,7 @@ def merge_pot(from_filenames, to_filename):
         if msgcat is None:
             msgcat = "/usr/bin/msgcat"
-        cmdline = [msgcat, "--use-first"]
+        cmdline = [msgcat, "--use-first", "-F"]
```
```suggestion
        cmdline = [msgcat, "-F"]
```
We need to cumulate file positions

freakboy3742 reviewed (changes requested):
One code style tweak, and one request for clarification - it’s entirely possible you’re correct in what you’ve done, but I’m not sufficiently familiar with msgcat in practice to be confident in that.

johnzhou721 commented:
Um... what's the tweak? Forgot to finish review?

freakboy3742 reviewed (changes requested):
Hrm… not sure what happened there - I must have neglected to confirm the comment I wrote.

The two comments were:

1. Using long form flags - using `—sort-by-file` instead of `-F`
2. Can you confirm why this is the right approach? I’m not an expert on msgcat (and I can’t confirm anything manually right now), but `—use-first` suggests merging keys, which seems preferable to duplication

johnzhou721 commented:
@freakboy3742 Quite the opposite. Merge everything *except* keys -- but that also means that the "header" is being merged into something like
```
# #-#-#-#-#  templates-72gdmk0l.pot (PROJECT VERSION)  #-#-#-#-#
# Translations template for PROJECT.
# Copyright (C) 2025 ORGANIZATION
# This file is distributed under the same license as the PROJECT project.
# FIRST AUTHOR <EMAIL@ADDRESS>, 2025.
#
#, fuzzy
msgid ""
msgstr ""
"#-#-#-#-#  contents.pot (PACKAGE VERSION)  #-#-#-#-#\n"
"Project-Id-Version: PACKAGE VERSION\n"
"Report-Msgid-Bugs-To: \n"
"POT-Creation-Date: 2025-06-02 08:23+AWST\n"
"PO-Revision-Date: YEAR-MO-DA HO:MI+ZONE\n"
"Last-Translator: FULL NAME <EMAIL@ADDRESS>\n"
"Language-Team: en <LL@li.org>\n"
"Language: en\n"
"MIME-Version: 1.0\n"
"Content-Type: text/plain; charset=UTF-8\n"
"Content-Transfer-Encoding: 8bit\n"
"#-#-#-#-#  templates-72gdmk0l.pot (PROJECT VERSION)  #-#-#-#-#\n"
"Project-Id-Version: PROJECT VERSION\n"
"Report-Msgid-Bugs-To: EMAIL@ADDRESS\n"
"POT-Creation-Date: 2025-06-29 12:06-0500\n"
"PO-Revision-Date: YEAR-MO-DA HO:MI+ZONE\n"
"Last-Translator: FULL NAME <EMAIL@ADDRESS>\n"
"Language-Team: LANGUAGE <LL@li.org>\n"
"MIME-Version: 1.0\n"
"Content-Type: text/plain; charset=utf-8\n"
"Content-Transfer-Encoding: 8bit\n"
"Generated-By: Babel 2.17.0\n"
```
Which is problematic. It's keeping both versions.

**CONCLUSION** hold this PR. just found this issue

johnzhou721 commented:
That said, xgettext is the canoncial way to merge pots, so I'm trying that.

johnzhou721 commented:
These concerns should now be resolved.

johnzhou721 commented on CHANGELOG.md:
```diff
@@ -1,5 +1,10 @@
+## 0.5.5
+
+* Ensure that POT content is now sorted by path when merging POTs from multiple sources (i.e., templates and content).
+* POT contents now cumulate (i.e., instead of `--use-first`) when merging POTs using `msgcat`, such that source information is preserved from all POT files.
```
```suggestion
* `xgettext` is used to merge POT files instead of `msgcat`, providing a better header and merging of same strings from different sources.
```

johnzhou721 commented:
DO NOT MERGE. It changes the creation date of the POT unnessacraily.

johnzhou721 commented on lektor_i18n.py:
```diff
@@ -137,14 +137,28 @@ def write_pot(self, pot_filename, language):
+        # Inject the creation date back into the produced file
+        with open(to_filename, 'r', encoding='utf-8') as f:
+            finishedfile_orig = f.read()
+        replacement = r'\1' + date1 + r'\3'
```
```suggestion
        replacement = r'\g<1>' + date1 + r'\g<3>'
```

johnzhou721 commented:
There. This works now.

This is what's called Yak Shaving I guess... just to have all the source places merged together, I removed an option from msgcat and realized xgettext is the canoncial way (https://www.gnu.org/software/gettext/manual/html_node/msgcat-Invocation.html, "To concatenate POT files, better use xgettext, not msgcat, because msgcat would choke on the undefined charsets in the specified POT files.") and can handle this better...

johnzhou721 commented:
Again, as mentioned on the other thread, I really apologize for shaving all those yaks.

johnzhou721 commented:
I've added ngettext, pgettext as available functions, and fixed the limitation about non-English content language, and fixed a bug where when we're updating translations, the English po file is not updated with the strings. A demo for how the first bug doesn't apply anymore can be found at 4480606. That is generated by setting the default language to Farsi and deleting both the Farsi and English translation files, updating the plugin, and lektor-building.

johnzhou721 commented on CHANGELOG.md:
```diff
@@ -1,5 +1,13 @@
+* `xgettext` is used to merge POT files instead of `msgcat`, providing a better header and merging of same strings from different sources.
+* The initially generated PO files will now have a header compatible with GNOME's Translation Editor, since they will have a non-placeholder `Project-Id-Version`. Existing users hitting this problem will need to fill in the `Project-Id-Version` header manually.
+* Translations in templates now provide pgettext and ngettext methods.
+* The bug with non-English content is resolved.
```
```suggestion
* The bug where deletion of strings from the English PO file with non-English content is resolved.
* When updating translated PO files, the content-language PO file strings are automatically filled with the message IDs.
```

johnzhou721 commented:
Hold on... DO NOT MERGE THIS

When filling in strings for source language, we need to not fill when there already is a string, and document the fact that my heuristic for where to fill in msgid_plural is bad and whenever that is done I need to mark it fuzzy

johnzhou721 commented on lektor_i18n.py:
```diff
@@ -159,6 +182,28 @@ def parse_templates(to_filename):
+                entry.msgstr_plural[idx] = ''
+    po.save(save_path or po_filepath)
+
+def fill_translations(po_filepath, save_path=None):
+    po = polib.pofile(po_filepath)
+    for entry in po:
+        entry.msgstr = entry.msgid
+        if entry.msgstr_plural:
+            for idx in entry.msgstr_plural:
+                if int(idx) == 0:
+                    entry.msgstr_plural[idx] = entry.msgid
+                else:
+                    entry.msgstr_plural[idx] = entry.msgid_plural
+    po.save(save_path or po_filepath)
```
```suggestion
def fill_translations(po_filepath, save_path=None):
    po = polib.pofile(po_filepath)
    
    for entry in po:
        if not entry.msgstr:
            entry.msgstr = entry.msgid

        plural_updated = False
        if entry.msgstr_plural:
            for idx in entry.msgstr_plural:
                if not entry.msgstr_plural[idx]:
                    entry.msgstr_plural[idx] = entry.msgid if int(idx) == 0 else entry.msgid_plural
                    plural_updated = True

        if plural_updated and 'fuzzy' not in entry.flags:
            entry.flags.append('fuzzy')

    po.save(save_path or po_filepath)
```

johnzhou721 commented:
Hmm... in the bugfix I used the msgstr number to determine whether the singular form or the plural form needs to get filled in... it's only a good heuristic for some languages, so I documented that and if it's not English I mark the auto-filled plurals as fuzzy. See the changelog/readme for more details.

(FYI i patched to fill msgstr with msgids automatcially on the source language po file and clear all the other POs after initial msginit so the bug listed on the README is finally resolved, but here comes plural handling etc etc)

johnzhou721 commented on lektor_i18n.py:
```diff
@@ -159,6 +182,35 @@ def parse_templates(to_filename):
+                entry.msgstr_plural[idx] = ''
+    po.save(save_path or po_filepath)
+
+def fill_translations(po_filepath, save_path=None):
+    po = polib.pofile(po_filepath)
+    
+    for entry in po:
+        if not entry.msgstr:
+            entry.msgstr = entry.msgid
+
+        plural_updated = False
+        if entry.msgstr_plural:
+            for idx in entry.msgstr_plural:
+                if not entry.msgstr_plural[idx]:
+                    entry.msgstr_plural[idx] = entry.msgid if int(idx) == 0 else entry.msgid_plural
+                    plural_updated = True
+
+        if plural_updated and 'fuzzy' not in entry.flags and 'en' not in basename(po_filepath):
+            entry.flags.append('fuzzy')
+
+    po.save(save_path or po_filepath)
```
```suggestion
def fill_translations(po_filepath, save_path=None):
    po = polib.pofile(po_filepath)
    
    for entry in po:
        if not entry.msgstr:
            entry.msgstr = entry.msgid

        need_plural_fill = False
        if entry.msgstr_plural:
            for idx in entry.msgstr_plural:
                if not entry.msgstr_plural[idx]:
                    need_plural_fill = True

        if need_plural_fill and '+en.po' in basename(po_filepath):
            for idx in entry.msgstr_plural:
                if not entry.msgstr_plural[idx]:
                    need_plural_fill = True
    po.save(save_path or po_filepath)
```

johnzhou721 commented on README.md:
```diff
@@ -166,6 +165,14 @@ All translation files (`contents-*.po`) are then compiled and merged with the or
 
+### Plural Forms
+
+If you're using `{% pluralize %}` or `ngettext` or the like in your Jinja templates, make sure you fill in the plural forms in the PO headers manually, then make sure you have the correct
+number of `msgstr[x]`s.  The plugin automatically fills `msgstr`s into the PO file of your source lanaguage (which msginit only does for English), but since it doesn't parse plural forms,
+care must be taken to ensure that there are proper plural form `msgstr`s in your source language PO file; in particular, for languages that does not simply use `msgstr[0]` as singular and
+`msgstr[1]` as plural, you must manually alter your PO file to have the correct plurals.  Therefore, all such filled-in source-language plural PO entries that aren't in English are marked
+fuzzy, and the plural forms would need to be adjusted manually.
```
```suggestion
If you're using `{% pluralize %}` or `ngettext` or the like in your Jinja templates, make sure you fill in the plural forms in the PO headers manually, then make sure you have the correct
number of `msgstr[x]`s.  The plugin automatically fills `msgstr`s into the PO file of your source lanaguage (which msginit only does for English), but since it doesn't parse plural forms,
any non-English PO file will not have its plural message strings filled in.  Those must be done manually in the source-language PO file if simply singular and plural strings does not suffice .
```

johnzhou721 commented on lektor_i18n.py:
```diff
@@ -567,3 +625,6 @@ def on_after_build_all(self, builder, **extra):
             po_file.generate()
+            if language == self.content_language:
+                fill_translations(os.path.join(po_file.i18npath, po_file.FILENAME_PATTERN.format(po_file.language)))
+            po_file.reformat()
```
```suggestion
                po_file.reformat()
```

johnzhou721 commented on lektor_i18n.py:
```diff
@@ -186,6 +238,7 @@ def _msg_init(self):
         ]
         reporter.report_debug_info("msginit cmd line", cmdline)
         portable_popen(cmdline, cwd=self.i18npath).wait()
+        clear_translations(os.path.join(self.i18npath, self.FILENAME_PATTERN.format(self.language)))
```
```suggestion
        clear_translations(os.path.join(self.i18npath, self.FILENAME_PATTERN.format(self.language)))
        self.reformat()
```

johnzhou721 reviewed (commented):
Not sure... the diff is still very large and rewraps a bunch of chinese simplified strings on the beeware.github.io PR even when using ubuntu 24.04... I'm trying to see if this is the trouble.

johnzhou721 commented on lektor_i18n.py:
```diff
@@ -159,6 +182,54 @@ def parse_templates(to_filename):
+                else:
+                    plurals_empty = False
+
+        if need_plural_fill and '+en.po' in basename(po_filepath):
+            for idx in entry.msgstr_plural:
+                if not entry.msgstr_plural[idx]:
+                    entry.msgstr_plural[idx] = entry.msgid if int(idx) == 0 else entry.msgid_plural
```
```suggestion
        if need_plural_fill and '+en.po' in basename(po_filepath):
            for idx in entry.msgstr_plural:
                entry.msgstr_plural[idx] = entry.msgid if int(idx) == 0 else entry.msgid_plural
```

johnzhou721 commented:
latest commit has a logic error, I will refactor by extracting clearing entry into seperate functions and clearing entry when fill translation entry is fuzzy, this is a note to self.

kattni commented:
Hey, John. I'm taking over the final review on this PR. I may be asking for clarification on your changes in this process.

My first question is, do you consider this ready for a final review? You repeatedly requested throughout the process that we hold off on merging it, so I want to ensure it's actually ready at this point before moving forward.

johnzhou721 commented:
@kattni Yes, please!  It's ready for final review -- I've made a lot of random changes here and there just to *completely* fixup this plugin so we hopefully don't need another update again.  If you need me to reexplain anything, let me know, as I was quite vague in communication when I started this PR.  I apoglogize for the noise.

kattni commented:
@johnzhou721 I would appreciate it if you can explain to me how you're testing these changes. I'd like to test it before we get into explanations.

johnzhou721 commented:
A lot of the testing happens at the beeware.github.io PR where lots of those changes are relevant and applied.

> * Ensure that POT content is now sorted by path when merging POTs from multiple sources (i.e., templates and content).

You can see that the new POT at https://github.com/beeware/beeware.github.io/pull/684/files#diff-c8f80bf8f257ddef4811618539fadecda3407d93671f14c95b81e3a161dc2c1c is sorted properly.  There's lots of diffs in that file though because of the next change quoted below, however that does make the POT context formatting consistent with the PO files.

> * `xgettext` is used to merge POT files instead of `msgcat`, providing a better header and merging of same strings from different sources. This is used to ensure that all context will be kept; removing —use-first from msgcat simply causes header trouble.

Merging of different context is demonstrated at https://github.com/beeware/beeware.github.io/pull/684/files#diff-c8f80bf8f257ddef4811618539fadecda3407d93671f14c95b81e3a161dc2c1cR5102-R5103 -- --use-first of msgcat seems to just use the first context of those strings; however if we remove that flag, the headers are different and the header msgstr will be a complete mess.

> * The initially generated PO files will now have a header compatible with GNOME's Translation Editor, since they will have a non-placeholder `Project-Id-Version`. This is mostly for my local work — I’ve filled the project id versions manually into the existing POs — but have the good effect of adding a license header (Same license as BeeWare) on top of newly generated POs; I have not went in and added these would’ve-been-generated headers on the existing PO files, though.

Use ``lektor quickstart`` to generate a project, and config lektor-i18n-plugin -- add babel.cfg and config/i18n.ini, along with adding the dependency and alternatives into the lektorproject file.  Then go to models, pick something random and make it translatable = True

Now lektor build and the POT will be
```
# SOME DESCRIPTIVE TITLE.
# Copyright (C) YEAR THE PACKAGE'S COPYRIGHT HOLDER
# This file is distributed under the same license as the transtest package.
# FIRST AUTHOR <EMAIL@ADDRESS>, YEAR.
#
#, fuzzy
msgid ""
msgstr ""
"Project-Id-Version: transtest 1.0\n"
"Report-Msgid-Bugs-To: \n"
"POT-Creation-Date: 2025-10-11 09:15+CST\n"
"PO-Revision-Date: YEAR-MO-DA HO:MI+ZONE\n"
"Last-Translator: FULL NAME <EMAIL@ADDRESS>\n"
"Language-Team: LANGUAGE <LL@li.org>\n"
"Language: \n"
"MIME-Version: 1.0\n"
"Content-Type: text/plain; charset=UTF-8\n"
"Content-Transfer-Encoding: 8bit\n"

#: (content/blog/contents+en.lr:blog.title) https://website_url/blog/
msgid "Blog"
msgstr ""
```

showing that Project-Id-Version is properly filled in.

> * Translations in templates now provide pgettext and npgettext methods. ~~The pgettext is for the sprint helping string (pluralized with speaker count) and the gold member_s_ on the front page (pluralization, desperate bug I fixed);~~ pgettext is used to give more context and to work around a bug where trailing spaces get trimmed if regular gettext is used.

OK I've made a mistake in the listing here.  Ignore the sentence I've striked through here.

pgettext is used for member badges.  If the preview of the jinja i18n translation PR renders Katie's "badge" at /zh_CN/about/team/ ``超能力：Batavia、网站、高级养蜂师`` with no extra spaces, it should be working since pgettext is used here to translate the item separators into these Chinese variants.

> * The limitation where deletion of strings from the English PO file with non-English content is required is resolved. See the deletion in the README.
> * When updating translated PO files, the content-language PO file strings are automatically filled with the message IDs. Plural forms aren’t filled unless it’s English. [EDIT] This is actually a functional issue; when adding a plural form to a previously singular-only string, msgmerging will actually fill the plural form with the singular form, but fuzzied!! The additional handling ensures that plurals are filled correctly for source-language POs in this case.

I'll come up with tests for these 2 a bit later

> * The first bug in [Two Seemingly Untranslated Strings beeware.github.io#689](https://github.com/beeware/beeware.github.io/issues/689) has been fixed (button on frontpage).

See preview for this.  The missing strings are described at beeware.github.io#689 and they should be there.

johnzhou721 commented:
> > * The limitation where deletion of strings from the English PO file with non-English content is required is resolved. See the deletion in the README.

Use `lektor quickstart` to generate a project, and config lektor-i18n-plugin -- add babel.cfg and config/i18n.ini [AS DESCRIBED IN THE README FILE], along with adding the dependency and alternatives into the lektorproject file. Then go to models, pick something random and make it translatable = True -- but this time make sure French is the primary language in the lektorproject and config/i18n.ini.  Build the project and find that the en PO file is empty, which means the limitation where strings must be manually deleted from the English PO file with non-english content is resolved.

> > * When updating translated PO files, the content-language PO file strings are automatically filled with the message IDs. Plural forms aren’t filled unless it’s English. [EDIT] This is actually a functional issue; when adding a plural form to a previously singular-only string, msgmerging will actually fill the plural form with the singular form, but fuzzied!! The additional handling ensures that plurals are filled correctly for source-language POs in this case.

Use `lektor quickstart` to generate a project, and config lektor-i18n-plugin -- add babel.cfg and config/i18n.ini [AS DESCRIBED IN THE README FILE], along with adding the dependency and alternatives into the lektorproject file. Then go to models, pick something random and make it translatable = True -- and then lektor build.  Now change the code for the navigation bar in layout.html in templates to say
```
        {% for href, title in [
          ['/blog', 'Blog'],
          ['/projects', _("Projects")],
          ['/about', 'About']
        ] %}
```

So now rerun lektor build and we find that the English PO file automatically populates the new string for Projects -- this doesn't happen in the old version of the plugin, the new Projects string will just stay blank:
```
# English translations for transtest package.
# Copyright (C) 2025 THE transtest'S COPYRIGHT HOLDER
# This file is distributed under the same license as the transtest package.
# Automatically generated, 2025.
#
msgid ""
msgstr ""
"Project-Id-Version: transtest 1.0\n"
"Report-Msgid-Bugs-To: \n"
"POT-Creation-Date: 2025-10-11 09:52+CST\n"
"PO-Revision-Date: 2025-10-11 09:52+CST\n"
"Last-Translator: Automatically generated\n"
"Language-Team: none\n"
"Language: en\n"
"MIME-Version: 1.0\n"
"Content-Type: text/plain; charset=UTF-8\n"
"Content-Transfer-Encoding: 8bit\n"
"Plural-Forms: nplurals=2; plural=(n != 1);\n"

#: (content/blog/contents+en.lr:blog.title) https://website_url/blog/
msgid "Blog"
msgstr "Blog"

#: templates/layout.html:14
msgid "Projects"
msgstr "Projects"
```

Now we translate the string in the French version (*ce n'est pas une traduction*, though) and save the file:
```
#: templates/layout.html:14
msgid "Projects"
msgstr "THIS IS A FRENCH PLURAL TRANSLATION"
```

Now, we introduce a pluralized version of the string -- we do this in the navigation bar in layout.html:
```
        {% for href, title in [
          ['/blog', 'Blog'],
          ['/projects', ngettext("Project", "Projects", 1)],
          ['/about', 'About']
        ] %}
```

Notice now the English translation is automatically filled correctly, showing that the code for autofilling new source-language translations handles plurals properly:
```
# English translations for transtest package.
# Copyright (C) 2025 THE transtest'S COPYRIGHT HOLDER
# This file is distributed under the same license as the transtest package.
# Automatically generated, 2025.
#
msgid ""
msgstr ""
"Project-Id-Version: transtest 1.0\n"
"Report-Msgid-Bugs-To: \n"
"POT-Creation-Date: 2025-10-11 09:52+CST\n"
"PO-Revision-Date: 2025-10-11 09:52+CST\n"
"Last-Translator: Automatically generated\n"
"Language-Team: none\n"
"Language: en\n"
"MIME-Version: 1.0\n"
"Content-Type: text/plain; charset=UTF-8\n"
"Content-Transfer-Encoding: 8bit\n"
"Plural-Forms: nplurals=2; plural=(n != 1);\n"

#: (content/blog/contents+en.lr:blog.title) https://website_url/blog/
msgid "Blog"
msgstr "Blog"

#: templates/layout.html:14
msgid "Project"
msgid_plural "Projects"
msgstr[0] "Project"
msgstr[1] "Projects"

#~ msgid "Projects"
#~ msgstr "Projects"
```

If the same test is performed on French with the source language, the new Projects string will not be automatically filled due to the lack in our ability to parse Plural-Forms to figure out what msgstr index to fill with msgid source and what to fill with plural msgid source.

kattni commented:
As noted [here](https://github.com/beeware/beeware.github.io/pull/684#issuecomment-3482946528), we're putting a pin in this for now, and will pick it up again if needed.

johnzhou721 commented:
Closing this since the new MkDocs rebuild is going *very* far along by now.  I will, though, independently maintain a copy of this plugin, as the fixes contained here are extensive and fixes some core assumptions/bugs in the program.

@kattni Unrelated -- I've been tracking the new website rebuild -- it looks really good so far, and very visually clean.  But I do have some minor suggestions about a couple of places.  Would it be too premature to file them now on beeware/website?  I don't want to interfere with y'all still refining it.

freakboy3742 commented:
We're not *quite* ready for an external review yet, but we're getting close. Keep an eye on beeware/website#4; as that checklist gets close to completion, we'll be looking for feedback.

johnzhou721 commented:
@freakboy3742 Thank you for letting me know.
